### ISSUE TITLE
Large Environment Properties

### ISSUE THREAD

**Issue by Robadob** [FLAMEGPU/FLAMEGPU2#162]
Striuvad has the requirement of storing ~32,000 host managed agents. These should be represented as environment properties (for technical reasons), but clearly won't fit in constant cache so need to be allocated in global memory (accessed via ldg?).

cc @ptheywood

**Comment by ptheywood** [comment]
The upper limit on the population is more like `2^12`, although in practice that will never be achieved.

**Comment by MILeach** [comment]
Came up with this which allows pretty uniform creation of properties with any dimensions [https://gist.github.com/MILeach/1ff2da85091705bdaec97b05504df5c4](link). I think this solves what you wanted with same class name with different template args @Robadob? Internally uses std::array at the moment to store the dimensions but could use any representation. Also currently uses c++17 for the fold expression but can workaround that if necessary

I'm in favour of allowing multiple dimensions as it gives a better abstraction of domain problems and reduces scope for error with type safety & bounds checking. It also is less work for a modeller if we do it, although even for multi-dimensional properties, some conversion will probably be required from location to index.

**Comment by Robadob** [comment]
Few problems with your suggested code @MILeach :
* Using `new` in device code is VERY expensive.
* RTC is unable to compile `std::array` currently (the system header is missing, it would probably work if a replacement was added to Jitify).

**Comment by MILeach** [comment]
Yeah the new and std::array were really just there as examples/placeholders, memory allocation would obviously need to be handled differently. 

Yeah that would work too, I think it depends where we want to draw the line on flexibility. My implementation would allow 4D ones too which could be used to track different layers across a 3D environment etc in a single macro property.

**Comment by Robadob** [comment]
Yeah, I agree yours is far more dynamic, which is always a good thing. I'm just not really 100% up on variadic template soup.

4D makes sense to me in a context of a 3D array of Array variables, so 3D, 1D, rather than outright 4D. E.g. if  a user needed a 3D array of vec3.

**Comment by MILeach** [comment]
Yes exactly re: 4D, alternatively an array of field variables, e.g. if you had 100 different chemical concentrations or something they could be indexed by the first value, then x, y, z.

**Comment by ptheywood** [comment]
> Going to be working on what we've coined 'macro environment' over the next few weeks. 
 Which will combine this and #308.

Is this going to be the only way to have atomically mutable environmental properties? In which case I'm not convinved by the "MacroProperty" name (but I don't have a better suggestion yet).
Macro property doesn't really fit to strit's use case either IMO, but that might just be bias towards my use of macro in the past. 


> mondus:  I would have no objection to simplifying this greatly and having only scalar and 1d array values


I'm not as strongly opposed to providing 2d/3d versions, *but* for strit's use case which this issue originally mentions, the environment is not a square / cube, but a hexagonal lattice representing the inner surface of a torus, which can be mapped to 2D square indexing with some extra steps, but I'd be inclined to just use 1D and then perform the hex coord mapping directly to that (which has to be done for agents anyway, so the methods are already in place)

The only downside I see otherwise of treating them all the same is any overhead costs at runtime when not using all 3 dims (i.e. a single value). Probably negligable compard to curve itself though / the cost of the atomic ops, and as you say with constexpr if's this shoulnd't actually be a concern, so discard this point.

> Also currently uses c++17 for the fold expression but can workaround that if necessary

We will be going c++17 when viable. Bede is testing the RHEL8/CUDA 11.3 install, so we can sortof claim that's supported (although it may become unavailable for a short while). Bessemer is still 10.x only, I've nudged on github, but given it's August things might be slow (and the use of easybuild might prevent very recent 11.x, though <= 11.1 should be achievable). I believe the driver is already 11.0, but the cuda install wasn't done / made public? 

> ignoring them at compile time via constexpr

`if constexpr` is c++17 too, though optimised builds of regular if statements which only contain constexpr'd values might get optimised out (I assume?), which would/should apply to the template args. Debug builds might still have the dead branches kicking around, but then who cares. 


Otherwise it seems reasonable to me / Matt's rough implementation seems sane. 

## tldr;
+ Looks decent. 
+ We will be able to rely on c++17 feature "soon"
    + Until then CI will be unhappy for CUDA 10 builds.
+ Not convinved by the name, `macro` has other implications to me. I don't have a better name though.

**Comment by Robadob** [comment]
> Is this going to be the only way to have atomically mutable environmental properties? In which case I'm not convinved by the "MacroProperty" name (but I don't have a better suggestion yet).

Yeah I'm not dead set on the name either. I would like to have a separate form validation only atomics, which are only active if built with a compile macro on. But that's a different topic.

Rest is fair

**Comment by ptheywood** [comment]
> Yeah I'm not dead set on the name either. I would like to have a separate form validation only atomics, which are only active if built with a compile macro on. But that's a different topic.

I think that will be able to be done already using the above with enough macro soup by the user, but a separate implementaion in the library which uses constexpr's (rather than adding more macro's, when we're c++17) might be a nice addition

**Comment by Robadob** [comment]
Better to be consistent and have all our build flags as macros, than to confuse it by mixing and matching.

**Comment by Robadob** [comment]
Closed by #643
